extensions/amp-stream-gallery/1.0/stream-gallery.js

```diff
@@ -0,0 +1,223 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {BaseCarousel} from '../../amp-base-carousel/1.0/base-carousel';
+import {setStyle} from '../../../src/style';
+import {
+  useCallback,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './stream-gallery.jss';
+
+const DEFAULT_VISIBLE_COUNT = 1;
+const OUTSET_ARROWS_WIDTH = 100;
+
+/**
+ * @param {!StreamGalleryDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function StreamGallery({
+  arrowPrev: customArrowPrev,
+  arrowNext: customArrowNext,
+  children,
+  className,
+  extraSpace,
+  insetArrowVisibility,
+  loop,
+  maxItemWidth = Number.MAX_VALUE,
+  minItemWidth = 1,
+  maxVisibleCount = Number.MAX_VALUE,
+  minVisibleCount = 1,
+  outsetArrows,
+  peek = 0,
+  snap,
+  ...rest
+}) {
+  const classes = useStyles();
+  const ref = useRef(null);
+  const [visibleCount, setVisibleCount] = useState(DEFAULT_VISIBLE_COUNT);
+  const arrowPrev = useMemo(
+    () =>
+      customArrowPrev ?? <DefaultArrow by={-1} outsetArrows={outsetArrows} />,
+    [customArrowPrev, outsetArrows]
+  );
+  const arrowNext = useMemo(
+    () =>
+      customArrowNext ?? <DefaultArrow by={1} outsetArrows={outsetArrows} />,
+    [customArrowNext, outsetArrows]
+  );
+
+  const measure = useCallback(
+    (containerWidth) =>
+      getVisibleCount(
+        maxItemWidth,
+        minItemWidth,
+        maxVisibleCount,
+        minVisibleCount,
+        children.length,
+        outsetArrows,
+        peek,
+        containerWidth,
+        ref.current.node
+      ),
+    [
+      maxItemWidth,
+      minItemWidth,
+      maxVisibleCount,
+      minVisibleCount,
+      children.length,
+      outsetArrows,
+      peek,
+    ]
+  );
+
+  // Adjust visible slide count when container size or parameters change.
+  useLayoutEffect(() => {
+    if (!ref.current) {
+      return;
+    }
+    const node = ref.current.root;
+    if (!node) {
+      return;
+    }
+    // Use local window.
+    const win = node.ownerDocument.defaultView;
+    const observer = new win.ResizeObserver((entries) => {
+      const last = entries[entries.length - 1];
+      setVisibleCount(measure(last.contentRect.width));
+    });
+    observer.observe(node);
+    return () => observer.disconnect();
+  }, [measure]);
+
+  return (
+    <BaseCarousel
+      advanceCount={Math.floor(visibleCount)}
+      arrowPrev={arrowPrev}
+      arrowNext={arrowNext}
+      className={`${className ?? ''} ${classes.gallery} ${classes[extraSpace]}`}
```

Comment:
`classes[extraSpace]` won't work, b/c `classes` is an enum and thus we can only access it using the "known" names.

Reply:
In my manual testing this works -- isn't `"around"` known as the property name whose value is the generated class name? This is what I see when I toggle `extraSpace` on (`== "around"`) and off (border applied to see the space available):
![image](https://user-images.githubusercontent.com/10456171/97363713-5fa9b900-1879-11eb-855d-0c31a8534c2b.png)
![image](https://user-images.githubusercontent.com/10456171/97363725-620c1300-1879-11eb-806d-3f3bd12c19c7.png)

Reply:
It works for now. But we're trying to completely optimize enums where they are always replaced with constants. I.e. you can think of `classes` being mostly gone at some point.

Reply:
Oooh I see. Okay, replaced this with the class ternary.

Reply:
LG